src/main/kotlin/uk/gov/justice/digital/hmpps/crimeportalgateway/application/SoapHeaderAddressInterceptor.kt

```diff
@@ -22,29 +28,39 @@ class SoapHeaderAddressInterceptor : EndpointInterceptor {
 
         val soapResponseHeader = soapResponseMessage.saajMessage?.soapPart?.envelope?.header
         if (soapResponseHeader != null) {
-            val requestHeaders: Map<String, String>? = getHeaderMap(soapRequestMessage.saajMessage?.soapPart?.envelope?.header)
-            addTextNodeToNewElement(soapResponseHeader, "Action", requestHeaders?.get("Action")
+            val requestHeaders: Map<String, String> = getHeaderMap(soapRequestMessage.saajMessage?.soapPart?.envelope?.header)
+            addTextNodeToNewElement(soapResponseHeader, "Action", requestHeaders["Action"]
                     ?: "externalDocument")
             addTextNodeToNewElement(soapResponseHeader, "MessageID", UUID.randomUUID().toString())
-            addTextNodeToNewElement(soapResponseHeader, "To", requestHeaders?.get("From")
+            addTextNodeToNewElement(soapResponseHeader, "To", requestHeaders["From"]
                     ?: "")
-            addTextNodeToNewElement(soapResponseHeader, "RelatesTo", requestHeaders?.get("MessageID")
+            addTextNodeToNewElement(soapResponseHeader, "RelatesTo", requestHeaders["MessageID"]
                     ?: "")
 
             val soapElement = soapResponseHeader.addChildElement("From", "", SOAP_ENV_ADDRESS_NS)
-            addTextNodeToNewElement(soapElement, "Address", requestHeaders?.get("To") ?: "")
+            addTextNodeToNewElement(soapElement, "Address", requestHeaders["To"] ?: "")
         }
 
         return true
     }
 
-    override fun handleFault(messageContext: MessageContext?, p1: Any?): Boolean {
-        TODO("Not yet implemented")
+    override fun handleFault(messageContext: MessageContext, p1: Any?): Boolean {
+        telemetryService.trackEvent(TelemetryEventType.COURT_LIST_MESSAGE_ERROR)
+        try {
+            val buffer = ByteArrayOutputStream()
+            messageContext.response.writeTo(buffer)
+            val payload: String = buffer.toString(StandardCharsets.UTF_8.name())
+            log.error(payload)
+        } catch (e: IOException) {
+            throw object : WebServiceClientException("Can not write the SOAP fault into the out stream", e) {
```

Comment:
I wonder if catching and rethrowing gives any value here, it would be nice for readability if you didn't catch the IOException at all

Reply:
Yes. Tested with SOAP fault and it's all OK